src/twisted/persisted/dirdbm.py

```diff
@@ -220,99 +247,102 @@ def setdefault(self, key, value):
         @param key: The key to lookup
-        
+
         @param default: The value to return if the given key does not exist
-        
+
         @return: The value associated with C{key} or C{default} if not
-        C{self.has_key(key)}
+        L{DirDBM.has_key(key)}
         """
         if key in self:
             return self[key]
         else:
             return default
 
+
     def __contains__(self, key):
         """
-        C{key in dirdbm}
-
-        @type key: str
-        @param key: The key to test
-                
-        @return: A true value if C{self.has_key(key)}, a false value otherwise.
+        L{DirDBM.has_key}
         """
-        assert type(key) == str, "DirDBM key must be a string"
-        key = self._encode(key)
-        return os.path.isfile(os.path.join(self.dname, key))
+        return self.has_key(key)
+
 
     def update(self, dict):
         """
-        Add all the key/value pairs in C{dict} to this dirdbm.  Any conflicting
-        keys will be overwritten with the values from C{dict}.
+        Add all the key/value pairs in L{dict} to this dirdbm.  Any conflicting
+        keys will be overwritten with the values from L{dict}.
 
         @type dict: mapping
         @param dict: A mapping of key/value pairs to add to this dirdbm.
         """
         for key, val in dict.items():
             self[key]=val
-            
+
+
     def copyTo(self, path):
         """
         Copy the contents of this dirdbm to the dirdbm at C{path}.
-        
-        @type path: C{str}
+
+        @type path: L{str}
         @param path: The path of the dirdbm to copy to.  If a dirdbm
         exists at the destination path, it is cleared first.
-        
+
         @rtype: C{DirDBM}
         @return: The dirdbm this dirdbm was copied to.
         """
-        path = os.path.abspath(path)
-        assert path != self.dname
-        
-        d = self.__class__(path)
+        path = FilePath(path)
+        assert path != self._dnamePath
+
+        d = self.__class__(path.path)
         d.clear()
         for k in self.keys():
             d[k] = self[k]
         return d
 
+
     def clear(self):
         """
         Delete all key/value pairs in this dirdbm.
         """
         for k in self.keys():
             del self[k]
 
+
     def close(self):
         """
         Close this dbm: no-op, for dbm-style interface compliance.
         """
 
+
     def getModificationTime(self, key):
         """
         Returns modification time of an entry.
-        
+
         @return: Last modification date (seconds since epoch) of entry C{key}
         @raise KeyError: Raised when there is no such key
         """
-        assert type(key) == str, "DirDBM key must be a string"
-        path = os.path.join(self.dname, self._encode(key))
-        if os.path.isfile(path):
-            return os.path.getmtime(path)
+        if not type(key) == bytes:
+            raise TypeError("DirDBM key must be bytes")
```

Comment:
Does `AssertionError` ➜ `TypeError` fit with the compatibility policy?
IIRC not really, though I strongly agree that this is an improvement.